File: java/rocksjni/external_sst_file_info.cc
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree. An additional grant
// of patent rights can be found in the PATENTS file in the same directory.
//
// This file implements the "bridge" between Java and C++ and enables
// calling c++ rocksdb::BackupableDB and rocksdb::BackupableDBOptions methods
// from Java side.

#include <jni.h>

#include "include/org_rocksdb_ExternalSstFileInfo.h"
#include "rocksdb/sst_file_writer.h"

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    newExternalSstFileInfo
 * Signature: ()J
 */
jlong Java_org_rocksdb_ExternalSstFileInfo_newExternalSstFileInfo__(
    JNIEnv *env, jclass jcls) {
  return reinterpret_cast<jlong>(new rocksdb::ExternalSstFileInfo());
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    newExternalSstFileInfo
 * Signature: (Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;JJII)J
 */
jlong Java_org_rocksdb_ExternalSstFileInfo_newExternalSstFileInfo__Ljava_lang_String_2Ljava_lang_String_2Ljava_lang_String_2JJII(
    JNIEnv *env, jclass jcls, jstring jfile_path, jstring jsmallest_key,
    jstring jlargest_key, jlong jsequence_number, jlong jfile_size,
    jint jnum_entries, jint jversion) {
  const char *file_path = env->GetStringUTFChars(jfile_path, NULL);
  const char *smallest_key = env->GetStringUTFChars(jsmallest_key, NULL);
  const char *largest_key = env->GetStringUTFChars(jlargest_key, NULL);
  auto *external_sst_file_info = new rocksdb::ExternalSstFileInfo(
      file_path, smallest_key, largest_key,
      static_cast<rocksdb::SequenceNumber>(jsequence_number),
      static_cast<uint64_t>(jfile_size), static_cast<int32_t>(jnum_entries),
      static_cast<int32_t>(jversion));
  env->ReleaseStringUTFChars(jfile_path, file_path);
  env->ReleaseStringUTFChars(jsmallest_key, smallest_key);
  env->ReleaseStringUTFChars(jlargest_key, largest_key);
  return reinterpret_cast<jlong>(external_sst_file_info);
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    setFilePath
 * Signature: (JLjava/lang/String;)V
 */
void Java_org_rocksdb_ExternalSstFileInfo_setFilePath(JNIEnv *env, jobject jobj,
                                                      jlong jhandle,
                                                      jstring jfile_path) {
  auto *external_sst_file_info =
      reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
  const char *file_path = env->GetStringUTFChars(jfile_path, NULL);
  external_sst_file_info->file_path = file_path;
  env->ReleaseStringUTFChars(jfile_path, file_path);
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    filePath
 * Signature: (J)Ljava/lang/String;
 */
jstring Java_org_rocksdb_ExternalSstFileInfo_filePath(JNIEnv *env, jobject jobj,
                                                      jlong jhandle) {
  auto *external_sst_file_info =
      reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
  return env->NewStringUTF(external_sst_file_info->file_path.data());
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    setSmallestKey
 * Signature: (JLjava/lang/String;)V
 */
void Java_org_rocksdb_ExternalSstFileInfo_setSmallestKey(
    JNIEnv *env, jobject jobj, jlong jhandle, jstring jsmallest_key) {
  auto *external_sst_file_info =
      reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
  const char *smallest_key = env->GetStringUTFChars(jsmallest_key, NULL);
  external_sst_file_info->smallest_key = smallest_key;
  env->ReleaseStringUTFChars(jsmallest_key, smallest_key);
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    smallestKey
 * Signature: (J)Ljava/lang/String;
 */
jstring Java_org_rocksdb_ExternalSstFileInfo_smallestKey(JNIEnv *env,
                                                         jobject jobj,
                                                         jlong jhandle) {
  auto *external_sst_file_info =
      reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
  return env->NewStringUTF(external_sst_file_info->smallest_key.data());
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    setLargestKey
 * Signature: (JLjava/lang/String;)V
 */
void Java_org_rocksdb_ExternalSstFileInfo_setLargestKey(JNIEnv *env,
                                                        jobject jobj,
                                                        jlong jhandle,
                                                        jstring jlargest_key) {
  auto *external_sst_file_info =
      reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
  const char *largest_key = env->GetStringUTFChars(jlargest_key, NULL);
  external_sst_file_info->largest_key = largest_key;
  env->ReleaseStringUTFChars(jlargest_key, largest_key);
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    largestKey
 * Signature: (J)Ljava/lang/String;
 */
jstring Java_org_rocksdb_ExternalSstFileInfo_largestKey(JNIEnv *env,
                                                        jobject jobj,
                                                        jlong jhandle) {
  auto *external_sst_file_info =
      reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
  return env->NewStringUTF(external_sst_file_info->largest_key.data());
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    setSequenceNumber
 * Signature: (JJ)V
 */
void Java_org_rocksdb_ExternalSstFileInfo_setSequenceNumber(
    JNIEnv *env, jobject jobj, jlong jhandle, jlong jsequence_number) {
  auto *external_sst_file_info =
      reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
  external_sst_file_info->sequence_number =
      static_cast<rocksdb::SequenceNumber>(jsequence_number);
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    sequenceNumber
 * Signature: (J)J
 */
jlong Java_org_rocksdb_ExternalSstFileInfo_sequenceNumber(JNIEnv *env,
                                                          jobject jobj,
                                                          jlong jhandle) {
  auto *external_sst_file_info =
      reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
  return static_cast<jlong>(external_sst_file_info->sequence_number);
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    setFileSize
 * Signature: (JJ)V
 */
void Java_org_rocksdb_ExternalSstFileInfo_setFileSize(JNIEnv *env, jobject jobj,
                                                      jlong jhandle,
                                                      jlong jfile_size) {
  auto *external_sst_file_info =
      reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
  external_sst_file_info->file_size = static_cast<uint64_t>(jfile_size);
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    fileSize
 * Signature: (J)J
 */
jlong Java_org_rocksdb_ExternalSstFileInfo_fileSize(JNIEnv *env, jobject jobj,
                                                    jlong jhandle) {
  auto *external_sst_file_info =
      reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
  return static_cast<jlong>(external_sst_file_info->file_size);
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    setNumEntries
 * Signature: (JI)V
 */
void Java_org_rocksdb_ExternalSstFileInfo_setNumEntries(JNIEnv *env,
                                                        jobject jobj,
                                                        jlong jhandle,
                                                        jint jnum_entries) {
  auto *external_sst_file_info =
      reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
  external_sst_file_info->num_entries = static_cast<uint64_t>(jnum_entries);
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    numEntries
 * Signature: (J)I
 */
jint Java_org_rocksdb_ExternalSstFileInfo_numEntries(JNIEnv *env, jobject jobj,
                                                     jlong jhandle) {
  auto *external_sst_file_info =
      reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
  return static_cast<jint>(external_sst_file_info->num_entries);
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    setVersion
 * Signature: (JI)V
 */
void Java_org_rocksdb_ExternalSstFileInfo_setVersion(JNIEnv *env, jobject jobj,
                                                     jlong jhandle,
                                                     jint jversion) {
  auto *external_sst_file_info =
      reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
  external_sst_file_info->version = static_cast<int32_t>(jversion);
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    version
 * Signature: (J)I
 */
jint Java_org_rocksdb_ExternalSstFileInfo_version(JNIEnv *env, jobject jobj,
                                                  jlong jhandle) {
  auto *external_sst_file_info =
      reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
  return static_cast<jint>(external_sst_file_info->version);
}

/*
 * Class:     org_rocksdb_ExternalSstFileInfo
 * Method:    disposeInternal
 * Signature: (J)V
 */
void Java_org_rocksdb_ExternalSstFileInfo_disposeInternal(JNIEnv *env,
                                                          jobject jobj,
                                                          jlong jhandle) {
  delete reinterpret_cast<rocksdb::ExternalSstFileInfo *>(jhandle);
}

File: java/src/main/java/org/rocksdb/ExternalSstFileInfo.java
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree. An additional grant
// of patent rights can be found in the PATENTS file in the same directory.

package org.rocksdb;

import java.io.File;

public class ExternalSstFileInfo extends RocksObject {
  public ExternalSstFileInfo() {
    super(newExternalSstFileInfo());
  }

  public ExternalSstFileInfo(final String filePath, final String smallestKey,
      final String largestKey, final long sequenceNumber, final long fileSize, final int numEntries,
      final int version) {
    super(newExternalSstFileInfo(ensureNotNullFilePath(filePath),
        ensureNotNullSmallestKey(smallestKey), ensureNotNullLargestKey(largestKey), sequenceNumber,
        fileSize, numEntries, version));
  }

  private static String ensureNotNullFilePath(final String filePath) {
    if (filePath == null) {
      throw new NullPointerException("filePath is null.");
    }
    return filePath;
  }

  private static String ensureNotNullSmallestKey(final String smallestKey) {
    if (smallestKey == null) {
      throw new NullPointerException("smallestKey is null.");
    }
    return smallestKey;
  }

  private static String ensureNotNullLargestKey(final String largestKey) {
    if (largestKey == null) {
      throw new NullPointerException("largestKey is null.");
    }
    return largestKey;
  }

  public void setFilePath(final String filePath) {
    setFilePath(nativeHandle_, filePath);
  }

  public String filePath() {
    return filePath(nativeHandle_);
  }

  public void setSmallestKey(final String smallestKey) {
    setSmallestKey(nativeHandle_, smallestKey);
  }

  public String smallestKey() {
    return smallestKey(nativeHandle_);
  }

  public void setLargestKey(final String largestKey) {
    setLargestKey(nativeHandle_, largestKey);
  }

  public String largestKey() {
    return largestKey(nativeHandle_);
  }

  public void setSequenceNumber(final long sequenceNumber) {
    setSequenceNumber(nativeHandle_, sequenceNumber);
  }

  public long sequenceNumber() {
    return sequenceNumber(nativeHandle_);
  }

  public void setFileSize(final long fileSize) {
    setFileSize(nativeHandle_, fileSize);
  }

  public long fileSize() {
    return fileSize(nativeHandle_);
  }

  public void setNumEntries(final int numEntries) {
    setNumEntries(nativeHandle_, numEntries);
  }

  public int numEntries() {
    return numEntries(nativeHandle_);
  }

  public void setVersion(final int version) {
    setVersion(nativeHandle_, version);
  }

  public int version() {
    return version(nativeHandle_);
  }

  private native static long newExternalSstFileInfo();

  private native static long newExternalSstFileInfo(final String filePath, final String smallestKey,
      final String largestKey, final long sequenceNumber, final long fileSize, final int numEntries,
      final int version);

  private native void setFilePath(final long handle, final String filePath);

  private native String filePath(final long handle);

  private native void setSmallestKey(final long handle, final String smallestKey);

  private native String smallestKey(final long handle);

  private native void setLargestKey(final long handle, final String largestKey);

  private native String largestKey(final long handle);

  private native void setSequenceNumber(final long handle, final long sequenceNumber);

  private native long sequenceNumber(final long handle);

  private native void setFileSize(final long handle, final long fileSize);

  private native long fileSize(final long handle);

  private native void setNumEntries(final long handle, final int numEntries);

  private native int numEntries(final long handle);

  private native void setVersion(final long handle, final int version);

  private native int version(final long handle);

  @Override protected final native void disposeInternal(final long handle);
}
